### PR TITLE
buildGoModule: allow modSha256 

### PR DESCRIPTION
**Pull request by zowoq**
I'm not really happy about proposing this but I think we need some way of reliably dealing with these cases when `go mod vendor` doesn't work that doesn't require manual intervention.

Allowing the use of the previous implementation as a last resort seems the easiest approach.

cc @kalbasit @marsam @Mic92

### CODE REVIEW

**Comment by kalbasit** [comment]
I don't have strong feelings either way tbh. This is an acceptable change as far as I am concerned.

File: pkgs/applications/networking/mailreaders/aerc/default.nix [hunk 1]
```diff
@@ -13,21 +13,8 @@ buildGoModule rec {
...
-  vendorSha256 = "1rqn36510m0yb7k4bvq2hgirr3z8a2h5xa7cq5mb84xsmhvf0g69";
-
-  overrideModAttrs = (_: {
-      postBuild = ''
-      cp -r --reflink=auto ${libvterm}/libvterm vendor/github.com/ddevault/go-libvterm
-      cp -r --reflink=auto ${libvterm}/encoding vendor/github.com/ddevault/go-libvterm
-      '';
-    });
+  vendorSha256 = null;
+  modSha256 = "0nnf873cx91jh7nvfdznf83s4g8sk9xliaia9vm0f5snd5s2r131";
```
Comment (Mic92) [inline pkgs/applications/networking/mailreaders/aerc/default.nix:17]: What is actually the reason for us having to copy stuff here? are those libraries out of data or is vendoring missing files?

**Comment by Mic92** [comment]
cc @c00w

**Comment by Mic92** [comment]
So this is basically the opposite of https://github.com/NixOS/nixpkgs/pull/93513

File: pkgs/applications/networking/mailreaders/aerc/default.nix (hunk 1 above)
Comment (zowoq) [reply pkgs/applications/networking/mailreaders/aerc/default.nix:17]: Vendoring missing files.

**Comment by zowoq** [comment]
> cc ...

We should see about getting a @NixOS/golang

> So this is basically the opposite of #93513

Yeah. 

I want to get rid of `modSha256` but until `runVend` or some other alternative works properly we may as well just use what we know works.

**Comment by zowoq** [comment]
With 20.09 getting closing this will at least buy us another few months to find a reliable solution.

**Review by c00w — CHANGES_REQUESTED** [review]
I think this is completely ridiculous.

@zowoq is the entire reason we haven't been able to merge runVend, so him claiming we're in some sort of helpless situation here seems extremely duplicitious.

Additionally, it's generally doable to fix specific issues with the vendor folder (as this code shows), in a fairly straightforward way, so I don't understand what benefit we get by undoing the go vendor work.

**Comment by Mic92** [comment]
@c00w I am not sure they are aware of you progress. I suggest @zowoq to have another look at your PR.

**Comment by Mic92** [comment]
I would be actually in favor of having vendoring to work with @c00w solution rather than having to maintain to different module implementation.

**Comment by zowoq** [comment]
Updated this to use https://github.com/goware/modvendor. 

The vendor command is a bit long but it seems to work.

File: pkgs/development/go-modules/generic/default.nix [hunk 2]
```diff
@@ -21,6 +21,8 @@
 # Whether to delete the vendor folder supplied with the source.
 , deleteVendor ? false
 
+, modvendorCopy ? false
```
Comment (Mic92) [inline pkgs/development/go-modules/generic/default.nix:24]: Can you mention this in `doc/languages-frameworks/go.xml`?

**Comment by Mic92** [comment]
What is the difference between this and `vend`?

**Comment by zowoq** [comment]
> Can you mention this in `doc/languages-frameworks/go.xml`?

I'll add something to the docs if we decide to go with this PR.

> What is the difference between this and `vend`?

It's a simpler tool. It doesn't wrap `go` commands, it just copies files.

Downside is that it is a simple tool and we need to tell it to what to copy. Also, if we ever need to add more file types for whatever reason we may end up have to fix hashes because if those new file types exist in current packages they weren't being copied. 

I'm not sure if this is better approach than `vend` or not, if we can get the `replace` problems fixed it may be better to stick with `vend`.

**Comment by Mic92** [comment]
> > Can you mention this in `doc/languages-frameworks/go.xml`?
> 
> I'll add something to the docs if we decide to go with this PR.
> 
> > What is the difference between this and `vend`?
> 
> It's a simpler tool. It doesn't wrap `go` commands, it just copies files.
> 
> Downside is that it is a simple tool and we need to tell it to what to copy. Also, if we ever need to add more file types for whatever reason we may end up having to fix hashes because if those new file types exist in current packages they weren't being copied.
> 
> I'm not sure if this is better approach than `vend` or not, if we can get the `replace` problems fixed it may be better to stick with `vend`.

Having an explicit list does not sounds like a good idea. However I am concerned that if we ever add any file to the whitelist, we might break hashes. Would it be a bad idea to just add everything?

**Comment by zowoq** [comment]
> Would it be a bad idea to just add everything?

We can add as many file types as we want but anything else would require patching.

File: pkgs/development/go-modules/generic/default.nix [hunk 3]
```diff
@@ -88,6 +90,9 @@ let
         exit 10
       fi
       go mod vendor
+      if ${ lib.boolToString modvendorCopy }; then
+        ${modvendor}/bin/modvendor -copy="**/*.c **/*.cc **/*.cpp **/*.cxx **/*.h **/*.hh **/*.hpp **/*.hxx **/*.inc **/*.m **/*.proto **/*.s **/*.S **/*.swig **/*.swigcxx **/*.sx **/*.syso" -v
+      fi
```
Comment (Mic92) [inline pkgs/development/go-modules/generic/default.nix:95]: Is this supported?

```suggestion
        ${modvendor}/bin/modvendor -copy="**/*" -v
      fi
```

Comment (zowoq) [reply pkgs/development/go-modules/generic/default.nix:95]: No.

**Pull request closed without merging**